File: algorithm/python/FindEventualSafeStates.py
class Solution:
    def eventualSafeNodes(self, graph):
      length = len(graph)
      color = [0 for _ in range(length)]
      return [i for i in range(length) if self.isSafety(i, color, graph)]
    
    def isSafety(self, node, color, graph):
      if color[node]:
        return color[node] == 2
      
      color[node] = 1
      for i in graph[node]:
          if not self.isSafety(i, color, graph):
            return False
      color[node] = 2
      return True

File: algorithm/JavaScript/FindEventualSafeStates.js
/**
 * @param {number[][]} graph
 * @return {number[]}
 */
 var eventualSafeNodes = function(graph) {
    
  const n = graph.length
  const color = new Array(n).fill(0)

  const dfs = (node) => {

    if (color[node] > 0) {
      return color[node] === 2
    }
    color[node] = 1
    for (const nextNode of graph[node]) {
      if (!dfs(nextNode)) return false
    }
    color[node] = 2
    return true

  }
  const res = []
  for (let i = 0; i < n; i++) {
    dfs(i) ? res.push(i) : null
  }
  return res

};